### DISCUSSION TITLE
Why does this offset fail?

### DISCUSSION THREAD

**Discussion by raydeleu** [sgenoud/replicad#22] [Q&A]
I am still experimenting with a curved shape, trying to apply an offset. However, the offset on this shape fails. I thought  it could be caused by the sharp edge. A small test shape of a house with a pointed roof however worked fine. What do I miss? 

![makeOffset-fail](https://user-images.githubusercontent.com/38007983/172644297-f231b861-0352-40e0-ad27-736a617de8fd.png)

I added the code below: 

```` javascript
const main = ({ Sketcher, 
                makeOffset }) => 

{
// create a house with a pointed roof
let houseSketch = new Sketcher("XZ")
        .vLine(50)
        .line(10,25)
        .line(10,-25)
        .vLine(-50)
        .close()

let house = houseSketch.extrude(40)

// offset this shape
house = makeOffset(house,3).translate([0,60,0])

// create a shape with a wave-like curve
let p0 = [0,0]
let p1 = [50,100]
let p2 = [60,-95]
let p3 = [80,30]
let p4 = [100,25]
let points = [p1,p2,p3,p4]

let testBezier = new Sketcher("XZ")
        .movePointerTo(p0)
        .bezierCurveTo(p4,points)
        .vLine(-30)
        .hLine(-100)
        .close()
testBezier = testBezier.extrude(30)

// offset this shape
// testBezier = makeOffset(testBezier,3)

house = house.fuse(testBezier)

return house
}
````

**Comment by sgenoud** [comment]
I am not sure how what happens here. The `makeOffset` function is a very small wrapper around opencascade's [BRepOffsetAPI_MakeOffsetShape](https://dev.opencascade.org/doc/refman/html/class_b_rep_offset_a_p_i___make_offset_shape.html#a173d06428d1cdeea1c4f38fd6f80043c).

```js
export const makeOffset = (
  face: Face,
  offset: number,
  tolerance = 1e-6
): Shape3D => {
  const oc = getOC();
  const offsetBuilder = new oc.BRepOffsetAPI_MakeOffsetShape_2(
    face.wrapped,
    offset,
    tolerance,
    oc.BRepOffset_Mode.BRepOffset_Skin as any,
    false,
    false,
    oc.GeomAbs_JoinType.GeomAbs_Arc as any,
    false
  );

  const newShape = cast(downcast(offsetBuilder.Shape()));
  offsetBuilder.delete();

  if (!isShape3D(newShape)) throw new Error("Could not offset to a 3d shape");
  return newShape;
};
```

You may want to play with the options that I do not expose.

The problem might be that:

> All the faces of the shape S should be based on the surfaces with continuity at least C1.

Normally the Bezier should be C1 though...

I don't have much time this week, but I will be happy to modify the function if you find what is going on!

**Reply by raydeleu** [reply]
I tried the same shape in CascadeStudio and there it is possible to apply an offset. 

![wave-offset-cs](https://user-images.githubusercontent.com/38007983/173131241-d6938a97-adc8-4cff-b714-1448ff1095fd.jpg)

I tried to copy the function `Offset` from CascadeStudio into my source code, but I quickly ran into the issue that this code requires other functions as well, leading me down the rabbitshole. Even after some experiments I was not able to get it running. 

Then I tried to copy the function you supplied above and tinker with the values and the calls. But also there I stumbled upon things I simply do not understand. Pasting the function into the source code and renaming it did not work, probably because the code is written to be used from a library. I did not succeed to rewrite it as a simple function. Furthermore I tried to analyse the parameters in your code by reading the documentation at https://ocjs.org. At the page https://ocjs.org/reference-docs/classes/BRepOffsetAPI_MakeOffset_2 I found a different declaration of the function, requiring only 3 parameters instead of 8. The function `BRepOffsetAPI_MakeOffsetShape_2` did not appear on https://ocjs.org (notice the difference between `MakeOffset_2` and `MakeOffsetShape_2`). After these experiments I gave up, I have to learn a lot about programming in Javascript it seems.

**Comment by raydeleu** [comment]
<img width="647" alt="offset-cubicBezier" src="https://user-images.githubusercontent.com/38007983/173139287-55fb0771-d65e-4f62-aa8e-b9abf1861229.PNG">

<img width="443" alt="offset-cubicBezier-1" src="https://user-images.githubusercontent.com/38007983/173139800-b71dbc80-8bad-43ad-a537-e2976298abef.PNG">

I had a small success by replacing the `bezierCurveTo` by a `cubicBezierCurveTo`. It is really amazing that this works , especially considering the almost overlapping bulges at the end of this random shape. 
With the `quadraticBezierCurveTo` it also works fine. Can we conclude that the error lies in the `bezierCurveTo` function? 

<img width="410" alt="offsetQuadratic" src="https://user-images.githubusercontent.com/38007983/173143039-b8c13913-e372-4d87-b6d4-6d73fc8a932f.png">
